### PR TITLE
Allow to skip a build on commit messages

### PR DESCRIPTION
**Pull request by stsewd**
This implements a feature discussed here #871 

When a webhook is called, its receive the last commit after the push event,
the following marks are searched in the commit message:

- [skip doc]
- [skip docs]
- [docs skip]
- [doc skip]

If one is found, the build process is skipped.

Still wip, this is what I have so far:

- [x] Change GitHub webhook
- [x] Change GitLab webhook
- [x] Change BitBucket webhook
- [x] Change Generic webhook (backwards compatible)
- [x] fix tests
- [x] Add tests
- [x] Update docs

### CODE REVIEW

File: readthedocs/restapi/views/integrations.py [hunk 1]
```diff
@@ -159,7 +167,13 @@ def handle_webhook(self):
         # Handle push events and trigger builds
         if event == GITHUB_PUSH:
             try:
-                branches = [self.data['ref'].replace('refs/heads/', '')]
+                branch = {
```
Comment (stsewd) [inline readthedocs/restapi/views/integrations.py]: I change branch to be a dict with `name` and `last_commit` as elements.
`last_commit` must have an `id` and `message`.

File: readthedocs/restapi/views/integrations.py [hunk 2]
```diff
@@ -191,7 +213,13 @@ def handle_webhook(self):
         # Handle push events and trigger builds
         if event == GITLAB_PUSH:
             try:
-                branches = [self.request.data['ref'].replace('refs/heads/', '')]
+                # GitLab only returns one branch.
+                branch = {
+                    'name': self.request.data['ref'].replace('refs/heads/', ''),
+                    # Assuming the first element is the last commit.
+                    'commit': self.request.data['commits'][0],
```
Comment (stsewd) [inline readthedocs/restapi/views/integrations.py]: Not sure about this, didn't much documentation about this on gitlab

File: readthedocs/restapi/views/integrations.py [hunk 3]
```diff
@@ -264,6 +320,19 @@ class APIWebhookView(WebhookMixin, APIView):
...
+        {
+            "branches": [{
+                "name": "branch-name",
+                "last_commit": {
+                    "id": "sha",
+                    "message": "Update README.md"
+                }
+            }]
+        }
+
```
Comment (stsewd) [inline readthedocs/restapi/views/integrations.py]: This is what I proposed for the general webhook, maintaining the backwards compatibility.

File: readthedocs/core/views/hooks.py [hunk 4]
```diff
@@ -75,9 +92,17 @@ def build_branches(project, branch_list):
...
+        versions = project.versions_from_branch_name(branch['name'])
         to_build = set()
         not_building = set()
+        if _contains_skip_mark(commit):
+            log.info('(Branch Build) Skip build %s', project.slug)
+            not_building = {
+                version.slug
+                for version in versions
+            }
+            return (to_build, not_building)
```
Comment (stsewd) [inline readthedocs/core/views/hooks.py]: Here is the logic for skipping the build, not sure if this is the correct place

File: readthedocs/restapi/views/integrations.py [hunk 5]
```diff
@@ -137,8 +137,16 @@ class GitHubWebhookView(WebhookMixin, APIView):
...
+            "head_commit": {
+                "id": "sha",
+                "message": "Update README.md",
+                ...
+            }
             ...
         }
+
+    See full payload here:
+    https://developer.github.com/v3/activity/events/types/#pushevent
```
Comment (stsewd) [inline readthedocs/restapi/views/integrations.py]: Are we using the v3 of the gh api?

File: readthedocs/rtd_tests/tests/test_integrations.py [hunk 6]
```diff
@@ -62,15 +71,17 @@ def test_exchange_form_request_body(self):
         integration = fixture.get(Integration, project=project,
                                   integration_type=Integration.GITHUB_WEBHOOK,
                                   provider_data='')
-        resp = client.post(
+        payload_decode = '{"head_commit": {"id": "3eea78b2", "message": "Update README.md"}, "ref": "exchange_form"}'
+        payload_encode = '%7B%22head_commit%22%3A%20%7B%22id%22%3A%20%223eea78b2%22%2C%20%22message%22%3A%20%22Update%20README.md%22%7D%2C%20%22ref%22%3A%20%22exchange_form%22%7D'
```
Comment (stsewd) [inline readthedocs/rtd_tests/tests/test_integrations.py]: I use this tool for encode/decode https://meyerweb.com/eric/tools/dencoder/

Note: the order matters!

**Comment by stsewd** [comment]
While updating some tests, I found that there is some old webhook code (https://github.com/rtfd/readthedocs.org/blob/master/readthedocs/core/views/hooks.py#L158) that is marked as deprecated, should we remove this code? (another PR) o should we still supported this v1 webhooks?

**Comment by humitos** [comment]
> that is marked as deprecated, should we remove this code? (another PR?) o should we still support the v1 webhooks?

Hehe, I asked the same to Eric and he checked the logs: many people still use this so we can't remove this endpoint :(

File: readthedocs/core/views/hooks.py [hunk 7]
```diff
@@ -66,6 +66,23 @@ def _build_version(project, slug, already_built=()):
         return None
 
 
+def _contains_skip_mark(commit):
```
Comment (humitos) [inline readthedocs/core/views/hooks.py]: Receive the message itself, instead of relying on a `commit` object.

File: readthedocs/core/views/hooks.py [hunk 8]
```diff
@@ -75,9 +92,17 @@ def build_branches(project, branch_list):
         not_building - a list of branches that we won't build
     """
     for branch in branch_list:
-        versions = project.versions_from_branch_name(branch)
+        commit = branch['last_commit']
+        versions = project.versions_from_branch_name(branch['name'])
         to_build = set()
         not_building = set()
+        if _contains_skip_mark(commit):
+            log.info('(Branch Build) Skip build %s', project.slug)
```
Comment (humitos) [inline readthedocs/core/views/hooks.py]: Add the reason of the skip, like "because contains skip mark" or something _better_ :)

File: readthedocs/core/views/hooks.py (hunk 4 above)
Comment (humitos) [reply readthedocs/core/views/hooks.py]: So, if the last commit contains the mark we don't build _any_ version?

I'm confused here since the next few lines there is a logic that can build some version but not some others comming in the same webhook, right? Shouldn't we apply a similar logic here also?

That `to_build` and `not_building` rely on the function `_build_version`. So, this logic shouldn't be in that method? (I don't know, but is has the logging integrated and returns what it needs. On the other hand, it seems that we shouldn't call that method if we already know that we need to skip the version but there are some logic to skip versions inside that function too :/

File: readthedocs/restapi/views/integrations.py (hunk 5 above)
Comment (humitos) [reply readthedocs/restapi/views/integrations.py]: It seems, https://developer.github.com/v3/#current-version, since we are not sending the `Accept` header and `v3` is the default.

File: readthedocs/core/views/hooks.py [hunk 9]
```diff
@@ -75,9 +92,17 @@ def build_branches(project, branch_list):
         not_building - a list of branches that we won't build
     """
     for branch in branch_list:
-        versions = project.versions_from_branch_name(branch)
+        commit = branch['last_commit']
+        versions = project.versions_from_branch_name(branch['name'])
```
Comment (humitos) [inline readthedocs/core/views/hooks.py]: Why did you change `branch` by `branch['name']` here?

File: readthedocs/restapi/views/integrations.py [hunk 10]
```diff
@@ -191,7 +212,13 @@ def handle_webhook(self):
         # Handle push events and trigger builds
         if event == GITLAB_PUSH:
             try:
-                branches = [self.request.data['ref'].replace('refs/heads/', '')]
+                # GitLab only returns one branch.
+                branch = {
+                    'name': self.request.data['ref'].replace('refs/heads/', ''),
```
Comment (humitos) [inline readthedocs/restapi/views/integrations.py]: Won't this interferes with your other PR at #3546?

File: readthedocs/restapi/views/integrations.py [hunk 11]
```diff
@@ -191,7 +212,13 @@ def handle_webhook(self):
...
+                    # Assuming the first element is the last commit.
+                    'last_commit': self.request.data['commits'][0],
+                }
+                branches = [branch]
```
Comment (humitos) [inline readthedocs/restapi/views/integrations.py]: Another thought here... If we already know at this point the last commit message, shouldn't skip here the call to `self.get_response_push` instead of going deeper in the path?

File: readthedocs/rtd_tests/tests/test_integrations.py (hunk 6 above)
Comment (humitos) [reply readthedocs/rtd_tests/tests/test_integrations.py]: Can you do this with a Python function to encode a JSON into an URL query string? (it should be there one in the standard library)

**Comment by humitos** [comment]
Also, if you can, please add test cases for this specific feature so we can be sure that it works and then we are safe while refactoring/restructuring the code.

File: readthedocs/core/views/hooks.py (hunk 4 above)
Comment (stsewd) [reply readthedocs/core/views/hooks.py]: > I'm confused here since the next few lines there is a logic that can build some version but not some others comming in the same webhook, right? Shouldn't we apply a similar logic here also?

Of what I understand, a branch/tag/commit can be tied to various versions so there is only one commit message for all versions, this means that all versions from this branch aren't going to be built. And also is about what you mention, we already know if the version isn't going to be built.

File: readthedocs/rtd_tests/tests/test_integrations.py (hunk 6 above)
Comment (stsewd) [reply readthedocs/rtd_tests/tests/test_integrations.py]: Sure, I think that's cleaner, just wanted to keep to the current tests style.

File: readthedocs/core/views/hooks.py (hunk 9 above)
Comment (stsewd) [reply readthedocs/core/views/hooks.py]: See https://github.com/rtfd/readthedocs.org/pull/3457#discussion_r158978235

File: readthedocs/restapi/views/integrations.py (hunk 10 above)
Comment (stsewd) [reply readthedocs/restapi/views/integrations.py]: Just when merging/rebasing, on the logic everything would work as expected :)

File: readthedocs/restapi/views/integrations.py (hunk 11 above)
Comment (stsewd) [reply readthedocs/restapi/views/integrations.py]: We need the (buid, not_building) from the `build_branches` function to make the response.

**Comment by stsewd** [comment]
@humitos thanks for the feedback, I have improve the log and added some tests, I also made sure to keep the changes backwards compatibility with the old webhooks and with the generic webhook.

**Comment by ericholscher** [comment]
Feels like this has a larger refactor in it. Is this still something useful that we should try and get in?

**Comment by stsewd** [comment]
> Feels like this has a larger refactor in it. Is this still something useful that we should try and get in?

Hmm, not sure. This will allow users to help us to save some resources, but I don't know how much.

**Comment by humitos** [comment]
This is "nice to have" but I'm not sure that it worth to add this complexity to the code. Also, I'm not sure how many people will use this feature in the end since it's something manually that also could "mess/dirty" the commit messages from the repo.

**Comment by humitos** [comment]
I'm closing this PR here because it was off our roadmap and we considering that adding this complexity to the code base for a feature that will affect more _our servers_ than the user itself does not worth it.

Thanks a lot for this code and I'm sorry that it's not going to be merged :disappointed:
